keps/sig-testing/20190605-presubmitconfig-in-tested-repo.md

```diff
@@ -0,0 +1,277 @@
+
+Currently, all jobs for Prow are configured in a central `infra-config` repository that is
+in most cases distinct from the repositories whose code is being tested in presubmits. This
+poses severall challenges:
+
+* It is not possible to see in a Pull Request that introduces a new presubmit if that presubmit will
+  actually pass. The two workarounds for this are to either have both the job author and the reviewer
+  use two CLIs to create a Pod from the job or to make the Job initially optional, verify it with a
+  test Pull Request to the code repository, then make it mandatory. Both of these workarounds are
+  cumbersome.
+* The same issues apply when doing changes to the config of a presubmit
+* When a project maintains multiple branches, e.G. because there are release branches, the
+  maintainers must remember to create a copy of the presubmit in the `infra-config` repository.
+  Otherwise it is easely possible that the presubmit becomes incompatible with one of the branches
+  it is used in, because somone makes a change to its config and forgets to test against all branches.
+  This is an additional step maintainers have to remember when branching off.
+
+
+### Goals
+
+* It is possible to version some or all presubmits of a given repository inside that repository in a
+  `yaml` file
+* This feature is optional and opt-in
+* The triggering of presubmits on pull request creation or updates continues to work and includes the
+  jobs that are managed inside the code repository
+* Re-Running tests via the `/retest` command continues to work and includes the jobs that are
+  managed inside the code repository
+* Explicitly executing optional tests via the `/test <<myjob>>` command continues to work and includes
+  all jobs that are managed inside the code repository
+* All the existing defaulting and validation for Presubmit jobs is being used to default and validate
+  jobs that are managed inside the code repository
+* If there is an error parsing, defaulting or validating the presubmits that are managed inside the
+  code repository, a comment will be posted to GitHub stating the error
+* Pull Requests on which an error occurred during parsing, defaulting or validation the presubmits that
+  are managed in the code repository are not consideres as merge candidates by Tide
+* Tide executes the presubmits that are defined inside the code repository when it re-tests
+* Renamed blocking presubmits added via pull request trigger a migration on all in-flight PRs
+* Removed blocking presubmits via pull request trigger a status retire on all in-flight PRs
+
+### Non-Goals
+
+* The option to configure Postsubmits or Periodics inside the tested repository. This may be
+  done in a future iteration.
+
+## Proposal
+
+It is proposed to introduce a the option to configure additional presubmits inside the code
+repositories that are tested by prow via a file named `prow.yaml`
+
+The minimum part of Prow this has to be integrated with is:
```

Comment:
There is a lot more than `tide` and `trigger` that needs to be updated. Any plugin or component that consumes job config will need to be updated otherwise they may be looking at a truncated list of jobs. This is problematic if a specific job is being looked up and cannot be found, but it is also problematic when doing operations across all jobs e.g. enforcing branch protection, determining if a context can be skipped, or trying to migrating a job's status context. If we don't know that a job is configured we may remove its branch protection or improperly allow it to be skipped.
The following are the components/plugins I can think of off the top of my head that handle job config and will most likely need to be updated. There may be more.
- update-config
- config-bootstrapper
- branchprotector
- status-reconciler
- tide
- trigger
- skip
- override
- refresh
- horologium
- gerrit
- crier
- sub
- mkpj
- check-config

Another thing to consider is that this will make loading the job config require fetching `prow.yaml` from every repo that configures ProwJobs. A simple `git clone` would make that extremely expensive so we'll probably need to do a shallow clone or download the `prow.yaml` file individually.  Allowing config to live outside the cluster will also mean that config needs to be reloaded whenever we want to check for changes and that config loading will now be susceptible to network problems.

I'm worried that making the current job configuration dynamic and dependent on the PR(s) in question is going to add a lot of complexity and inefficiency to components that need to iterate over the currently configured jobs.

Reply:
> There is a lot more than tide and trigger that needs to be updated. Any plugin or component that consumes job config will need to be updated otherwise they may be looking at a truncated list of jobs. This is problematic if a specific job is being looked up and cannot be found, but it is also problematic when doing operations across all jobs e.g. enforcing branch protection, determining if a context can be skipped, or trying to migrating a job's status context. If we don't know that a job is configured we may remove its branch protection or improperly allow it to be skipped.

Generally speaking I think the simplest approach for this problem would be to extend the `Config` type with to funcs for accessing `Presubmits` from it:

* `func (c *Config) GetPresubmits(gc *git.Client, org, repo, base string, heads ...string)([]Presubmit, error)` for components that have a revision they work on like `trigger`
* `func (c *Config) GetStaticPresubmits(org, repo string) []Presubmit` for components that do not have a revision they work on, e.G. the branchprotector

We could then move all components piece by piece over to use that and eventually unexport the `Presubmit` field from the `Config` type. WDYT?

> Another thing to consider is that this will make loading the job config require fetching prow.yaml from every repo that configures ProwJobs

There is the `prow/git` package which has caching, so this is quite efficient.

Re the components:

> * update-config

This doesn't access the `Presubmit` field and basically just copies the content of some files into a configmap. What do you think needs to be changed there?

> config-bootstrapper

Basically the same question as for update-confog

> branchprotector

Yeah, this will be a bit special. I don't think its feasible to apply branchprotection for jobs that are versioned in the `prow.yaml`, as a PR could remove them so for that PR, they must not be protected.

Instead,  I think it makes more sense to have one protected context `prow-config-parsing` to make sure that happens and do not have branchprotection for the jobs in `prow.yaml`

> status-reconciler

I don't think this needs any change. The only thing we need is handling for the situation that a base had a job that got removed. I've outlined an approach to solve that without the status-reconciler in this doc, because making the status-reconciler cope with a `prow.yaml` will be rather tricky as its edge-driven

> skip

This uses `trigger.FilterPresubmits` under the hood, so adjusting trigger will be sufficient

> override

This works on github status contexts only and doesn't access the Presubmit config

> refresh

This gets its data from Deck I guess? Ultimatively a `prowjobs.js` file

> horologium

This is just for periodics or not?

> gerrit

Thats a bit tricky as it doesn't use the `org/repo` format. What about simply saying this feature is not supported with Gerrit?

> crier

Crier works on the Prowjob objects only or not? From a quick glance  I can't see it accessing the ProwJobs from the config

> sub

Doesn't access any the Prowjobs from the config either

> mkpj

More or less the whole point of this is not to need mkpj, so I'd say we can skip that :)

> check-config

Sure we need to validate the config for this

I guess my point is: Lets just migrate everything that accesses the `ProwJobs` field from the config to instead use a getter and decide on a case-by-case basis what makes sense. We can do that before we actually add the feature, that way we will see if there are bigger problems with any components. WDYT?